### ISSUE TITLE
Trained model produces near-identical outputs for different inputs?

### ISSUE THREAD

**Issue by linhdvu14** [philipperemy/deep-speaker#13]
As title. I noticed the trained model produces near-identical embeddings for different input audios (even randomly generated values). Is anyone having this issue?

If it helps, I'm training with
- LibriSpeech dev-clean dataset: http://www.openslr.org/12/
- SAMPLE_RATE = 16000
- BATCH_NUM_TRIPLETS = 21
- TRUNCATE_SOUND_FIRST_SECONDS = 3
- Loss = 0.10225 when I stopped training

I've tried a few different datasets/settings and had the same results.

Probably related to #9 . Somehow the model decides the best way to go is make all embeddings the same, hence loss = margin...

**Comment by philipperemy** [comment]
Yeah the loss function was wrong.

it got updated a few minutes ago.

Should be fixed with https://github.com/philipperemy/deep-speaker/pull/12

Can you check?

**Comment by linhdvu14** [comment]
Thanks. With the latest code, training loss doesn't appear to converge even after 1100 iterations. Any pointers why that might be?

![untitled](https://user-images.githubusercontent.com/13968867/33978713-a297887c-e0d3-11e7-9f08-937f41ea30be.png)

**Comment by philipperemy** [comment]
@linhdvu14 yes I think we are truncating samples to only one second so it's impossible to make the difference between speakers. I guess the next step would be to use only 3 different speakers and try it again.

**Comment by linhdvu14** [comment]
Thanks, will do! Some (somewhat tangential) questions, if you don't mind:
- Deep Speaker section 3.1.1 uses batch normalization between each convolution and relu. Why do we only use BN within the ResBlocks, but not between conv[xx]-s layer and res[xx] layer?
- Triplets are currently randomly selected. Do you have plans to implement more efficient triplet mining algorithms?

**Comment by philipperemy** [comment]
@linhdvu14 

- I missed this point! Yes please update the model code :) 
- No plans for now. The goal is more to make it work nicely before implementing more efficient triplet mining algorithms.

Very happy and looking forward to seeing your future contributions!

**Comment by nathanquinteiro** [comment]
Is this repo still active? And has this problem been resolved?
I have run the model training with the latest code on my environment with GPUs and still observe the problem of near-identical embeddings for different inputs.

The loss converges to (BATCH_SIZE * alpha)

Thanks for your help, best regards, Nathan

**Comment by philipperemy** [comment]
@nathanquinteiro I haven't had much time to have a look at the repo recently.

I think to solve the problem of near-identical embeddings for different inputs, it's required to train using supervised learning (like they do in the paper), then apply the methodology used in this repo.

If you want to help, I'd be more than happy to merge your pull requests!

Thanks!

**Comment by suyuzhang** [comment]
@linhdvu14  @philipperemy  Why I change TRUNCATE_SOUND_SECONDS = (0.5, 3.5) always get this error: Error when checking model input?
I think this error is input the different shape. How can I fix it?
 
Thanks for your help, best regards, Suyu

**Comment by suyuzhang** [comment]
@linhdvu14  @philipperemy  Why I change TRUNCATE_SOUND_SECONDS = (0.5, 2.5) always get this error: Error when checking model input. I think this error is input shapes are different. How can I fix it?

Thanks for your help!  best regards, Suyu

**Comment by Curisan** [comment]
Hello. With the latest code, I implement hard negatives from the same batch for each iteration just as they did in the section 3.3. This seems to help convergence. 
![image_hard_negative](https://user-images.githubusercontent.com/22990773/38605897-2f04d776-3da7-11e8-9a80-41d8e44e045e.png)
As a comparison, the following figure show the convergence curve when triplets are randomly selected
![no_hard_negative](https://user-images.githubusercontent.com/22990773/38606045-9b95bafe-3da7-11e8-9081-0e7b55d03a38.png)
But the loss seems to converges to (BATCH_SIZE * alpha) as @nathanquinteiro mention. 
If it helps, I'm training with
LibriSpeech dev-clean dataset: http://www.openslr.org/12/
SAMPLE_RATE = 16000
BATCH_NUM_TRIPLETS = 8
TRUNCATE_SOUND_FIRST_SECONDS = 1
Loss ≈1.6 when I stopped training

**Comment by philipperemy** [comment]
@Curisan great I think we're moving to the right direction :)

**Comment by philipperemy** [comment]
@riddle-man yeah there's still some work to be done before it's fully functional. If you can help, It'll be highly appreciated :)

**Comment by WHATDOESTHEDEATHSAY** [comment]
@Curisan hello，I saw your answer on CSDN, there are some questions I would like to consult. The author of the project took 8 copies before and after the data processing part. What do you mean by 16? Thank you.

**Comment by philipperemy** [comment]
@riddle-man @linhdvu14  Try again with the new master. I pushed the latest code. CNN and everything and it seems to work great now :)
